src/coloring.rs

```diff
@@ -59,3 +67,209 @@ pub fn graph_greedy_color(py: Python, graph: &graph::PyGraph) -> PyResult<PyObje
     }
     Ok(out_dict.into())
 }
+
+struct MisraGriesAlgorithm<'a> {
+    graph: &'a StablePyGraph<Undirected>,
+    colors: DictMap<EdgeIndex, usize>,
+}
+
+impl<'a> MisraGriesAlgorithm<'a> {
+    pub fn new(graph: &'a StablePyGraph<Undirected>) -> Self {
+        let colors: DictMap<EdgeIndex, usize> = DictMap::new();
+        MisraGriesAlgorithm { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: NodeIndex) -> Vec<usize> {
+        let mut used_colors: Vec<usize> = Vec::new();
+        for edge in self.graph.edges(u) {
+            if let Some(color) = self.colors.get(&edge.id()) {
+                used_colors.push(*color);
+            }
+        }
+        used_colors
+    }
+
+    // Returns the smallest free (aka unused) color at node u
+    fn get_free_color(&self, u: NodeIndex) -> usize {
+        let used_colors = self.get_used_colors(u);
+        let mut c: usize = 0;
+        while used_colors.contains(&c) {
+            c += 1;
+        }
+        c
+    }
+
+    // Returns if color c is free at node u
+    fn is_free_color(&self, u: NodeIndex, c: usize) -> bool {
+        let used_colors = self.get_used_colors(u);
+        !used_colors.contains(&c)
+    }
+
+    // Returns the maximal fan on edge ee = (u, v) at u
+    fn get_maximal_fan(
+        &self,
+        ee: EdgeIndex,
+        u: NodeIndex,
+        v: NodeIndex,
+    ) -> Vec<(EdgeIndex, NodeIndex)> {
+        let mut fan: Vec<(EdgeIndex, NodeIndex)> = Vec::new();
+        fan.push((ee, v));
+
+        let mut neighbors: Vec<(EdgeIndex, NodeIndex)> = Vec::new();
+        for edge in self.graph.edges(u) {
+            neighbors.push((edge.id(), edge.target()));
+        }
```

Comment:
I know this is a theme now :) But I would use an iterator to construct this just so it can be done in one allocation. Something like:

```suggestion
        let mut neighbors: Vec<(EdgeIndex, NodeIndex)> = self.graph.edges(u)
            .map(|edge| (edge.id(), edge.target()))
            .collect();
```